app/utils/langchain-utils.js

```diff
@@ -28,7 +28,43 @@ const parseMessage = (req, message) => {
   }
 }
 
+const extractLinksForValidatingResponse = (jsonObj) => {
+  const entries = []
+
+  try {
+    // Function to check URLs and gather entries with valid URLs
+    const checkUrls = (obj, parent) => {
+      if (typeof obj === 'object' && obj !== null) {
```

Comment:
Instead of performing a recursive traverse, can't we just do `answer.items.map((item) => item.url)` and then map by url. Same for the context, `context.map((contextItem) => contextItem.pageContent)`

Reply:
Given that it runs per response once, the additional computation delay should be negligible but the upside is that it is scalable, independent of the format, and extracts the exact link from the string instead of returning a string.  it also retuns the object when the issue occurred instead of simply raising an error.
I am happy to undo the traversing if you want to make this a very specific function that relies on a hardcoded structure, but I feel that **Given that it runs per response once....**